### PR TITLE
Improve Startup Time by persisting more LSM Segment Information

### PR DESCRIPTION
**Pull request by etiennedi**
### What's being changed:

*tl;dr: This improves startup time with the Wikiepdia dataset from ~5min to ~1min. It does so by persisting things that we are currently unnecessarily computing at startup time.*

**Before**

image

**After**

image

## Long Form

Before this PR, whenever we would initiate an LSM store segment, we would have to do two operations: 

1. Build a bloom filter for the segment. This required parsing all the keys in the segment's index structure. It was thus a function of the segments size.
2. Calculate the "count net additions" (CNAs). CNAs are a way to identify how many objects are in a bucket quickly. Because LSM stores are append-only, an update and a delete are also "just a write" – so simply counting keys isn't enough. We need to know if a key is a new insertion (count + 1), an updated (count unchanged) or a delete of an existing key (count -1). 

Both of these operations can easily be persisted on disk as well. This eliminates the need to recalculate them. The Bloom filter depends purely on the segment's contents. Since segments are immutable, the bloom filter is also immutable. The CNA depends on the contents of the segments and the contents of all previous segments. Since new segments can only be appended at the end (i.e., after an existing segment), the CNA is unaffected. Segments before or can only be changed by merging two together ("compaction"); however, the merged segment must have the same contents as its two inputs, thus not affecting the CNA.

This statement is verified extensively through [a brand new stress-test pipeline](https://github.com/semi-technologies/weaviate-chaos-engineering/pull/23) that produces a lot of compactions on a dataset, including plenty of deletions. It validates the object count at many points, including before, during, and after compactions.

## Backward Compatibility

This change is fully backward-compatible. When upgrading to this version, the files cannot yet be read from the disk because they don't exist. Therefore in the init phase, they are created (in-mem) as they would have been before. However, once created, they are persisted on disk. On subsequent starts, the files are read from disk, thus speeding up startup considerably.

## Checksums

This PR introduces new files next to the `segment-[0-9]+.db` files with extensions `.bloom` for bloom filters, and `.cna` for count net additions. Both new files make use of checksums. If a checksum is invalid - for example because Weaviate crashed while partially writing a file or hardware data corruption - the files are discarded and rebuilt on the fly. 

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [x] Chaos pipeline run or not necessary. [Link to pipeline](https://app.travis-ci.com/github/semi-technologies/weaviate-chaos-engineering)
- [x] All new code is covered by tests where it is reasonable: Yes, the remaining uncovered lines are all `if err != nil` branches on mostly os/file level operations
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/lsmkv/segment_bloom_filters.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+	}
+
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("close bloom filter file: %w", err)
+	}
+
+	return nil
+}
+
+func (ind *segment) loadBloomFilterSecondaryFromDisk(pos int) error {
```
Comment (dirkkul) [inline adapters/repos/db/lsmkv/segment_bloom_filters.go:217]: The handling for primary and secondary seems to be almost identical, can you move the identical parts into extra methods?

File: adapters/repos/db/lsmkv/segment_bloom_filters_test.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	"testing"
+
+	"github.com/sirupsen/logrus/hooks/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCreateBloomOnFlush(t *testing.T) {
+	ctx := context.Background()
+	dirName := makeTestDir(t)
```
Comment (dirkkul) [inline adapters/repos/db/lsmkv/segment_bloom_filters_test.go:28]: replace with `t.TempDir()`?

File: adapters/repos/db/lsmkv/segment_bloom_filters_test.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+	files, err := os.ReadDir(dirName)
+	require.Nil(t, err)
+	_, ok := findFileWithExt(files, ".bloom")
+	require.True(t, ok)
+	_, ok = findFileWithExt(files, ".secondary.0.bloom")
+	require.True(t, ok)
+}
+
+func TestRepairCorruptedBloomOnInit(t *testing.T) {
```
Comment (dirkkul) [inline adapters/repos/db/lsmkv/segment_bloom_filters_test.go:99]: Could you go through the error cases that have no test coverage and see if you can hit a few of them with different corruptions/file removals etc?

File: adapters/repos/db/lsmkv/segment_net_count_additions.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+	if _, err := f.Write(data); err != nil {
+		return fmt.Errorf("write cna data to file: %w", err)
+	}
+
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("close cna file: %w", err)
+	}
+
+	return nil
+}
```
Comment (dirkkul) [inline adapters/repos/db/lsmkv/segment_net_count_additions.go:150]: This looks identical to the function for bloom filters. If it is not too much work, can you combine this?

**Review by dirkkul — APPROVED** [review]

File: adapters/repos/db/lsmkv/segment_net_count_additions.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+	data := buf.Bytes()
+	chksm := crc32.ChecksumIEEE(data)
+
+	f, err := os.Create(path)
+	if err != nil {
+		return fmt.Errorf("open file for writing: %w", err)
+	}
+
+	if err := binary.Write(f, binary.LittleEndian, chksm); err != nil {
```
Comment (redouan-rhazouani) [inline adapters/repos/db/lsmkv/segment_net_count_additions.go:137]: f.Close()

File: adapters/repos/db/lsmkv/segment_net_count_additions.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		return fmt.Errorf("write checkusm to file: %w", err)
+	}
+
+	if _, err := f.Write(data); err != nil {
+		return fmt.Errorf("write cna data to file: %w", err)
```
Comment (redouan-rhazouani) [inline adapters/repos/db/lsmkv/segment_net_count_additions.go:142]: f.Close()

File: adapters/repos/db/lsmkv/segment_net_count_additions.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+func (ind *segment) loadCountNetFromDisk() error {
+	f, err := os.Open(ind.countNetPath())
+	if err != nil {
+		return fmt.Errorf("open file for reading: %w", err)
+	}
+
+	buf := new(bytes.Buffer)
+	if _, err := buf.ReadFrom(f); err != nil {
+		return fmt.Errorf("read cna data from file: %w", err)
```
Comment (redouan-rhazouani) [inline adapters/repos/db/lsmkv/segment_net_count_additions.go:160]: f.Close()

File: adapters/repos/db/lsmkv/segment_net_count_additions.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	}
+	data := buf.Bytes()
+	if len(data) != 12 {
+		return ErrInvalidChecksum
+	}
+
+	chcksm := binary.LittleEndian.Uint32(data[:4])
+	actual := crc32.ChecksumIEEE(data[4:])
+	if chcksm != actual {
+		return ErrInvalidChecksum
```
Comment (redouan-rhazouani) [inline adapters/repos/db/lsmkv/segment_net_count_additions.go:170]: f.Close()

File: adapters/repos/db/lsmkv/segment_bloom_filters.go (hunk 1 above)
Comment (etiennedi) [reply adapters/repos/db/lsmkv/segment_bloom_filters.go:217]: Good point, let me check if there's a clean way of doing this.

File: adapters/repos/db/lsmkv/segment_bloom_filters_test.go (hunk 2 above)
Comment (etiennedi) [reply adapters/repos/db/lsmkv/segment_bloom_filters_test.go]: Done

**Review by redouan-rhazouani — CHANGES_REQUESTED** [review]

**Review by redouan-rhazouani — APPROVED** [review]

File: adapters/repos/db/lsmkv/segment_net_count_additions.go (hunk 4 above)
Comment (etiennedi) [reply adapters/repos/db/lsmkv/segment_net_count_additions.go:128]: Done

File: adapters/repos/db/lsmkv/segment_bloom_filters.go (hunk 1 above)
Comment (etiennedi) [reply adapters/repos/db/lsmkv/segment_bloom_filters.go]: I didn't combine the primary/secondary, but since I refactored the storeWithChecksum/loadWithChecksum parts, where I've already unified the common logic (your other comment) the remaining methods are now super short :-)
